riff-raff/app/deployment/actors.scala

```diff
@@ -115,209 +110,310 @@ class DeployMetricsActor extends Actor with Logging {
+  }
+  private def cleanup = {
+    finishRootContext()
+    deployCoordinator ! DeployCoordinator.CleanupDeploy(record.uuid)
+    context.stop(self)
+  }
+
+  override def toString: String = {
+    s"""
+       |UUID: ${record.uuid.toString}
+       |#Tasks: ${allDeployments.size}
+       |#Executing: ${executing.mkString("; ")}
+       |#Completed: ${completed.size} Failed: ${failed.size}
+       |#Done: ${completed.size+failed.size}
+     """.stripMargin
+  }
+
+  def receive = {
+    case Start() =>
+      try {
+        self ! ContextCreated(createContext)
+        self ! StartDeployment()
+      } catch {
+        case NonFatal(t) =>
+          failRootContext("Preparing deploy failed", t)
+          cleanup
+      }
+
+    case ContextCreated(preparedContext) =>
+      deployContext = Some(preparedContext)
+
+    case StartDeployment() =>
+      runDeployments(firstDeployments)
+
+    case DeploymentCompleted(deployment) =>
+      log.debug("Deployment completed")
+      markComplete(deployment)
+      nextDeployments(deployment) match {
+        case Deployments(deployments) =>
+          runDeployments(deployments)
+        case FinishPath() =>
+        case FinishDeploy() =>
+          cleanup
+      }
+
+    case DeploymentFailed(deployment, exception) =>
+      log.debug("Deployment failed")
+      markFailed(deployment)
+      if (isExecuting) {
+        log.debug("Failed during deployment but others still running - deferring clean up")
+      } else {
+        cleanup
+      }
+
+    case Terminated(actor) =>
+      failRootContext("DeploymentRunner unexpectedly terminated", new RuntimeException("DeploymentRunner unexpectedly terminated"))
+      log.warn(s"Received terminate from ${actor.path}")
+  }
+
+  private def createContext: DeployContext = {
+    DeployReporter.withFailureHandling(rootReporter) { implicit safeReporter =>
+      import Configuration.artifact.aws._
+      safeReporter.info("Reading deploy.json")
+      val s3Artifact = S3Artifact(record.parameters.build, bucketName)
+      val json = S3Artifact.withZipFallback(s3Artifact) { artifact =>
+        Try(artifact.deployObject.fetchContentAsString()(client).get)
+      }(client, safeReporter)
+      val project = JsonReader.parse(json, s3Artifact)
+      val context = record.parameters.toDeployContext(record.uuid, project, LookupSelector(), safeReporter, client)
+      if (context.tasks.toTaskList.isEmpty)
+        safeReporter.fail("No tasks were found to execute. Ensure the app(s) are in the list supported by this stage/host.")
+      context
+    }
+  }
+
+  private def runDeployments(deployments: List[DeploymentNode]) = {
+    log.debug(s"Running next deployments: $deployments")
+    try {
+      honourStopFlag(rootReporter) {
+        deployments.foreach { deployment =>
+          val deploymentRunner = context.watch(deploymentRunnerFactory(context, s"${record.uuid}-${deployment.pathName}-${deployment.priority}"))
```

Comment:
I don't think you need to call `context.watch` because you are the parent actor. You should automatically receive a `Terminate` message when the child dies.